R/methods_htest.R

```diff
@@ -662,235 +563,78 @@ model_parameters.svytable <- function(model, verbose = TRUE, ...) {
-  }
-
-  # reorder
-  col_order <- c(
-    "Chi2", "df", "df_error", "Cohens_g", "g", "Cohens_CI_low",
-    "Cohens_CI_high", "p", "Method", "method"
-  )
-  out <- out[col_order[col_order %in% names(out)]]
-  out
-}
-
-
-
-
-# effectsize Cohen's d / Hedges g (t-test) ----
-
-.add_effectsize_ttest <- function(model,
+.add_effectsize_htest <- function(model,
                                   out,
-                                  standardized_d = NULL,
-                                  hedges_g = NULL,
+                                  effectsize_type = NULL,
+                                  effectsize_adjust = TRUE,
                                   ci = 0.95,
                                   alternative = NULL,
-                                  verbose = TRUE,
-                                  ...) {
-  if (is.null(standardized_d) && is.null(hedges_g)) {
+                                  verbose = TRUE) {
+  # check if effect sizes are requested
+  if (!requireNamespace("effectsize", quietly = TRUE) || is.null(effectsize_type)) {
     return(out)
   }
 
-  if (requireNamespace("effectsize", quietly = TRUE)) {
-    # standardized d
-    if (!is.null(standardized_d)) {
-      es <- effectsize::effectsize(
-        model,
-        type = "cohens_d",
-        ci = ci,
-        alternative = alternative,
-        verbose = verbose,
-        ...
-      )
-      es$CI <- NULL
-      ci_cols <- grepl("^CI", names(es))
-      names(es)[ci_cols] <- paste0("d_", names(es)[ci_cols])
-      out <- cbind(out, es)
-    }
-
-    # Hedge's g
-    if (!is.null(hedges_g)) {
-      es <- effectsize::effectsize(
+  # try to extract effectsize
+  es <- tryCatch(
+    {
+      effectsize::effectsize(
         model,
-        type = "hedges_g",
+        type = effectsize_type,
         ci = ci,
         alternative = alternative,
-        verbose = verbose,
-        ...
+        adjust = effectsize_adjust,
+        verbose = verbose
       )
-      es$CI <- NULL
-      ci_cols <- grepl("^CI", names(es))
-      names(es)[ci_cols] <- paste0("g_", names(es)[ci_cols])
-      out <- cbind(out, es)
+    },
+    error = function(e) {
+      if (verbose) {
+        msg <- c(
+          paste0("Could not compute effectsize ", effectsize::get_effectsize_label(effectsize_type), "."),
+          paste0("Possible reason: ", e$message)
+        )
+        insight::format_alert(msg)
+      }
+      NULL
     }
-  }
-
-  # reorder
-  col_order <- c(
-    "Parameter1", "Parameter2", "Parameter", "Group",
-    "Mean_Parameter1", "Mean_Parameter2", "Mean_Group1", "Mean_Group2",
-    "mu", "Difference", "CI_low", "CI_high",
-    "t", "df_error",
-    "d", "Cohens_d", "d_CI_low", "d_CI_high",
-    "g", "Hedges_g", "g_CI_low", "g_CI_high",
-    "p", "Method", "method"
   )
+  # return if not successful
+  if (is.null(es)) {
+    return(out)
+  }
 
-  out <- out[col_order[col_order %in% names(out)]]
-  out
-}
-
-
+  ## TODO: check if effectsize prefixes are correct @mattansb
```

Comment:
@mattansb Can you please check if these prefixes are correct?

Reply:
Note sure what you mean here - should these have all the possible effect sizes?

Reply:
Yeah, I adapted this code from the BFBayesFactor method. I think we need this prefix for the correct CI columns? Since now all effectsizes are handled inside this method, we need to have all possible prefixes here, right?

Reply:
Can `effectsize::get_effectsize_name()` help somehow?

Reply:
Not sure how this is supposed to work?

``` r
effectsize::get_effectsize_name("eta")
#> character(0)
```

Reply:
``` r
npk.aov <- aov(yield ~ block + N * P, data = npk)

out <- effectsize::effectsize(npk.aov)
effectsize::get_effectsize_name(colnames(out))
#> [1] "Eta2_partial"

out <- effectsize::effectsize(npk.aov, partial = FALSE, type = "omega")
effectsize::get_effectsize_name(colnames(out))
#> [1] "Omega2"
```

<sup>Created on 2022-10-02 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Ok, this would mean that I can replace

```r
  prefix <- switch(effectsize_type,
    "cohens_g" = "Cohens_",
    "cramers_v" = "Cramers_",
    "phi" = "phi_",
    "cohens_d" = "d_",
    "hedges_g" = "g_",
    "rank_biserial" = "rank_biserial_",
    "rank_epsilon_squared" = "rank_epsilon_squared_",
    "kendalls_w" = "Kendalls_W_",
    "omega" = "Omega2_",
    "eta" = "Eta2_",
    "epsilon" = "Epsilon2_"
  )
  ```

  with

  ```r
  prefix <- paste0(effectsize::get_effectsize_name(colnames(es)), "_")
  ```
  ?

Reply:
Yes, I believe so (:

Reply:
My daughter would now say "To believe, you go to church." ;-)

Reply:
Here if you believe you go to synagogue 😉😂

Reply:
Ok, doesn't fully work. Because wie expect `"Cohens_d"` as column name for effectsize, but `"d_CI_low"` as column name for CI.

Reply:
That might depend on the input:
When data can be retrieved from htest is gives the full Cohen’s d:

``` r
T <- t.test(mtcars$mpg)
es <- effectsize::effectsize(T)
effectsize::get_effectsize_name(colnames(es))
#> [1] "Cohens_d"
```

But when data cannot be retrieved, it gives an approximation

``` r
T <- t.test(mpg ~ 1, data = mtcars)
es <- effectsize::effectsize(T)
#> Warning: Unable to retrieve data from htest object. Using t_to_d()
#> approximation.
effectsize::get_effectsize_name(colnames(es))
#> [1] "d"
```

<sup>Created on 2022-10-02 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Might be. But the initial implementation expects "long form" for effect size name, and "short form" for CIs (like `hedges_g` and `g_CI_low` and `g_CI_high`). I think all relevant effectsizes are covered now... 🤞